src/app/dataset-view/additional-components/metadata-component/components/add-push-source/add-push-source.component.ts

```diff
@@ -76,11 +71,61 @@ export class AddPushSourceComponent extends BaseMainEventComponent {
         }),
     });
 
-    public get readForm(): FormGroup {
+    public initEditForm(): void {
+        this.trackSubscription(
+            this.editService
+                .getEventAsYaml(this.getDatasetInfoFromUrl(), this.queryParamName ?? "")
+                .subscribe((result) => {
+                    this.history = this.editService.history;
+                    if (this.sourceNameNotExist()) {
+                        this.zone.run(() => {
+                            this.navigationServices.navigateToPageNotFound();
+                        });
+                    }
+                    if (result) {
+                        this.eventYamlByHash = result;
+                        const currentPushSourceEvent = this.editService.parseEventFromYaml(this.eventYamlByHash);
+                        this.addPushSourceForm.patchValue({ sourceName: currentPushSourceEvent.sourceName });
+                    }
+                    if (!this.queryParamName) {
+                        this.addPushSourceForm.controls.sourceName.addValidators(
+                            RxwebValidators.noneOf({
+                                matchValues: [...this.getAllSourceNames(this.history), "default"],
+                            }),
+                        );
+                    }
+                    this.cdr.detectChanges();
+                }),
+        );
+    }
+
+    private getAllSourceNames(historyUpdate: MaybeNull<DatasetHistoryUpdate>): string[] {
+        if (!historyUpdate) {
+            return [];
+        } else {
+            const data = historyUpdate.history
+                .filter((item: MetadataBlockFragment) => item.event.__typename === SupportedEvents.AddPushSource)
+                .map((data: MetadataBlockFragment) => data.event as AddPushSource);
+            return data.map((event) => event.sourceName ?? "");
+        }
+    }
+
+    private sourceNameNotExist(): boolean {
+        return (
+            !!this.queryParamName &&
+            !this.history?.history.filter(
+                (item) =>
+                    item.event.__typename === SupportedEvents.AddPushSource &&
+                    item.event.sourceName === this.queryParamName,
+            ).length
+        );
+    }
```

Comment:
Please invert the question to reduce complexity of the implementation. Make a method "sourceNameExists()" and invert it where it's called. Reading current double negations is really hard.

Reply:
Fixed